File: 2022/04.ts
import { range } from "./utils/array.ts";

function part1(pairs: number[][][]): number {
  let overlaps = 0;
  for (const pair of pairs) {
    const [a, b] = pair[0];
    const [c, d] = pair[1];
    if ((a <= c && d <= b) || (c <= a && b <= d)) {
      overlaps++;
    }
  }
  return overlaps;
}

function part2(pairs: number[][][]): number {
  let overlaps = 0;
  for (const pair of pairs) {
    for (let x = pair[0][0]; x <= pair[0][1]; x++) {
      if (range(pair[1][1] - pair[1][0] + 1, pair[1][0]).includes(x)) {
        overlaps++;
        break;
      }
    }
  }
  return overlaps;
}

const data = await Deno.readTextFile("./inputs/04.txt").then((data) => {
  return data
    .split("\n")
    .map((line) => line.split(","))
    .map((pairs) =>
      pairs.map((pair) => pair.split("-").map((num) => parseInt(num)))
    );
});

console.log(part1(data));
console.log(part2(data));

File: 2022/04.py
from utils import aoc

# type aliases
Bounds = tuple[int, int]
Pair = tuple[Bounds, Bounds]


def parse_range(range: str):
    start, end = map(int, range.split("-"))

    return start, end


def parse_input(input: list[str]) -> list[Pair]:
    pairs = []
    for line in input:
        ranges = line.split(",")
        pair = (parse_range(ranges[0]), parse_range(ranges[1]))
        pairs.append(pair)

    return pairs


def part_one(pairs: list[Pair]) -> int:
    overlaps = 0
    for pair in pairs:
        if (pair[0][0] <= pair[1][0] and pair[1][1] <= pair[0][1]) or (
            pair[1][0] <= pair[0][0] and pair[0][1] <= pair[1][1]
        ):
            overlaps += 1

    return overlaps


def part_two(pairs: list[Pair]) -> int:
    overlaps = 0
    for pair in pairs:
        if not (pair[0][1] < pair[1][0] or pair[1][1] < pair[0][0]):
            overlaps += 1

    return overlaps


if __name__ == "__main__":
    input = aoc.get_input("04")
    pairs = parse_input(input)
    print(part_one(pairs))
    print(part_two(pairs))
